code/core/src/main/java/com/adobe/marketing/mobile/ExtensionDiscovery.kt

```diff
@@ -0,0 +1,60 @@
+/*
+  Copyright 2024 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile
+
+import android.content.ComponentName
+import android.content.Context
+import android.content.pm.PackageManager
+import com.adobe.marketing.mobile.internal.CoreConstants
+import com.adobe.marketing.mobile.services.Log
+
+/**
+ * Responsible for discovering and retrieving a list of [Extension] classes
+ * that are bundled with the app.
+ */
+internal class ExtensionDiscovery {
+    private val SERVICE_NAME = "com.adobe.marketing.mobile.ExtensionDiscoveryService"
+    private val LOG_TAG = "ExtensionDiscovery"
+
+    fun getExtensions(context: Context): List<Class<out Extension>> {
```

Comment:
Reflection-based scanning can be resource-intensive; consider caching results in case this function is called multiple times by accident.

Reply:
Currently, we don't call it more than one. I'll add caching if that situation arises.